### PR TITLE
Issue #9149: fix stack overflow in Navigator.next

### PR DESCRIPTION
**Pull request by pbludov**
#issue #9149

We cannot avoid stack overflow in `Navigator.next`, so natural test for `InputMatchXpathNoStackoverflowError.java` replaced with synthetic test with 10K nested nodes.

### CODE REVIEW

**Review by rnveach — COMMENTED** [review]
> We cannot avoid stack overflow in Navigator.next

We can if we report issue to library creator and ask for them to make a fix.

https://github.com/checkstyle/checkstyle/pull/9158#pullrequestreview-562870946
https://github.com/checkstyle/checkstyle/issues/9149#issuecomment-754880447
I am still more for setting stack size to make CI more constant as previous PR had. Changing input file to a manual AST which amounts to the same thing does not do anything for us. We can just do an AST print of an input file and generate the AST the same way.

**Comment by rnveach** [comment]
@romani ping

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/TestUtil.java [hunk 1]
```diff
@@ -40,6 +44,15 @@
...
 public final class TestUtil {
 
+    /**
+     * The stack size used in {@link #executeWithLimitedStackSize(Executable)}.
+     * This value should be as small as possible.
+     *
+     * @see <a href="https://www.baeldung.com/jvm-configure-stack-sizes">
+     *      Configuring Stack Sizes in the JVM</a>
+     */
+    private static final int MINIMAL_STACK_SIZE = 144 << 10;
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/TestUtil.java:54]: please explain "144 << 10"
why "<<" is used ?
please provide more details on this ...

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/RootNodeTest.java [hunk 2]
```diff
@@ -423,4 +423,5 @@ public void testSameNodeInfo() {
         assertFalse(rootNode.isSameNodeInfo(null),
                 "Should return false, because object does not equal null");
     }
+
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/xpath/RootNodeTest.java:426]: revert.

**Review by romani — CHANGES_REQUESTED** [review]
items:

**Review by rnveach — CHANGES_REQUESTED** [review]
The commit name does not represent what this is, which is fixing tests so our own tests aren't likely to overflow.

I think this PR should be part of https://github.com/checkstyle/checkstyle/issues/9163 as it clearly represents changing tests and not the issue that is causing the overflow to begin with.

For this https://github.com/checkstyle/checkstyle/issues/9149 we should report the issue to saxon and have them try and fix the cause of the issue.

File: src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/TestUtil.java (hunk 1 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/internal/utils/TestUtil.java:54]: https://teamcity.jetbrains.com/viewLog.html?buildId=3291117&tab=Inspection&buildTypeId=Checkstyle_IdeaInspectionsPullRequest
image

I don't think that this inspection is useful. `144 * 1024` is clearly states that this is 144k. Should we ignore this IDEA inspection?

File: src/test/java/com/puppycrawl/tools/checkstyle/xpath/RootNodeTest.java (hunk 2 above)
Comment (pbludov) [reply src/test/java/com/puppycrawl/tools/checkstyle/xpath/RootNodeTest.java:426]: done

**Comment by pbludov** [comment]
PR moved to https://github.com/checkstyle/checkstyle/pull/9223

**Pull request closed without merging**